staging/src/k8s.io/client-go/features/features.go

```diff
@@ -0,0 +1,127 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package features
+
+import (
+	"sync/atomic"
+)
+
+// NOTE: types Feature, FeatureSpec, prerelease (and its values)
+// were duplicated from the component-base repository
+//
+// for more information please refer to https://docs.google.com/document/d/1g9BGCRw-7ucUxO6OtCWbb3lfzUGA_uU9178wLdXAIfs
+
+const (
+	// Values for PreRelease.
+	Alpha = prerelease("ALPHA")
+	Beta  = prerelease("BETA")
+	GA    = prerelease("")
+
+	// Deprecated
+	Deprecated = prerelease("DEPRECATED")
+)
+
+type prerelease string
+
+type Feature string
+
+type FeatureSpec struct {
+	// Default is the default enablement state for the feature
+	Default bool
+	// LockToDefault indicates that the feature is locked to its default and cannot be changed
+	LockToDefault bool
+	// PreRelease indicates the maturity level of the feature
+	PreRelease prerelease
+}
+
+// Reader indicates whether a given feature is enabled or not.
+type Reader interface {
+	// Enabled returns true if the key is enabled.
+	Enabled(key Feature) bool
+}
+
+// Registry represents an external feature gates registry.
+type Registry interface {
+	Add(map[Feature]FeatureSpec) error
+}
+
+// FeatureGates returns the feature gates exposed by this library.
+//
+// By default, only the default features gate will be returned.
+// The default implementation allows controlling the features
+// via environmental variables.
+// For example, if you have a feature named "MyFeature,"
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that the actual set of the features gates
+// might be overwritten by calling SetFeatureGates method.
+func FeatureGates() Reader {
+	return featureGates.Load().(*featureGatesWrapper).Reader
+}
+
+// AddFeaturesToExistingFeatureGates adds the default feature gates to the provided registry.
+// Usually this function is combined with SetFeatureGates to take control of the
+// features exposed by this library.
+func AddFeaturesToExistingFeatureGates(registry Registry) error {
+	return registry.Add(defaultKubernetesFeatureGates)
+}
+
+// SetFeatureGates overwrites the default implementation of the feature gate
+// used by this library.
+//
+// Useful for binaries that would like to have full control of the features
+// exposed by this library, such as allowing consumers of a binary
+// to interact with the features via a command line flag.
+//
+// For example:
+//
+//	// first, register client-go's features to your registry.
+//	clientgofeaturegate.AddFeaturesToExistingFeatureGates(utilfeature.DefaultMutableFeatureGate)
+//	// then replace client-go's feature gate implementation with your implementation
+//	clientgofeaturegate.SetFeatureGates(utilfeature.DefaultMutableFeatureGate)
+func SetFeatureGates(newFeatureGates Reader) {
```

Comment:
confusing that the interface is called Reader, but you set the feature gates. `type Gates interface` and `ReplaceGates` or so would be clearer.

`

Reply:
How about a mechanism to warn when the EnvVar variant is queried before this is called? That seems to be a bug users could experience that is super hard to track down.

Reply:
> confusing that the interface is called Reader, but you set the feature gates. type Gates interface and ReplaceGates or so would be clearer.

ok, will rename the interface to `Gates`.

Reply:
> How about a mechanism to warn when the EnvVar variant is queried before this is called? That seems to be a bug users could experience that is super hard to track down.

sure, i like it, I will open a new pr for this change.

Reply:
created https://github.com/kubernetes/kubernetes/pull/122641